subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -173,27 +190,34 @@ private String getFilename(Path dir) {
 
         @Override
         public FileVisitResult visitFile(Path file, @Nullable BasicFileAttributes attrs) {
-            String name = stringInterner.intern(file.getFileName().toString());
+            String name = intern(file.getFileName().toString());
             if (isAllowed(file, name, false, attrs, builder.getRelativePath())) {
-                if (attrs == null) {
-                    throw new FileSnapshottingException(String.format("Cannot read file '%s': not authorized.", file));
-                }
-                if (attrs.isSymbolicLink()) {
-                    // when FileVisitOption.FOLLOW_LINKS, we only get here when link couldn't be followed
-                    throw new FileSnapshottingException(String.format("Could not list contents of '%s'. Couldn't follow symbolic link.", file));
-                }
-                addFileSnapshot(file, name, attrs);
+                FileSystemLocationSnapshot snapshot = getSnapshot(file, attrs, name);
+                builder.visit(snapshot);
             }
             return FileVisitResult.CONTINUE;
         }
 
+        private FileSystemLocationSnapshot getSnapshot(Path file, BasicFileAttributes attrs, String name) {
+            try {
+                String absolutePath = intern(file.toString());
+                HashCode hash = hasher.hash(file.toFile(), attrs.size(), attrs.lastModifiedTime().toMillis());
+                FileMetadata fileMetadata = FileMetadata.from(attrs);
+                return new RegularFileSnapshot(absolutePath, name, hash, fileMetadata);
+            } catch (UncheckedIOException e) {
+                LOGGER.debug("Could not read file path '{}'.", file, e);
+                return getMissingSnapshot(file);
+            }
+        }
+
         @Override
         public FileVisitResult visitFileFailed(Path file, IOException exc) {
             // File loop exceptions are ignored. When we encounter a loop (via symbolic links), we continue
             // so we include all the other files apart from the loop.
             // This way, we include each file only once.
             if (isNotFileSystemLoopException(exc) && isAllowed(file, file.getFileName().toString(), false, null, builder.getRelativePath())) {
```

Comment:
`isAllowed` will return `false`, since `attrs == null`. So we will never add a missing file snapshot if visiting a file failed. I guess this is not expected. We should add a test to show that case.

Reply:
This is fixed in the latest version